### ISSUE TITLE
Mempool Lanes: introduce QoS to the mempool [tracking issue]

### ISSUE THREAD

**Issue by sergio-mena** [cometbft/cometbft#2803]
### Summary

In order to enable better management of transaction traffic at P2P level, we need to allow the app to classify transactions. For that, we introduce the concept of Quality of Service (QoS), well known in IP networks, which we realize as **mempool lanes**. [ADR-118](https://github.com/cometbft/cometbft/blob/main/docs/references/architecture/adr-118-mempool-lanes.md) describes this feature more in detail.

This issue will track:
* the discussion with our users
* design work
* implementation work
* any follow up

### Teams interested

1. Osmosis
2. Injective
3. dYdX
4. Astria

### Execution Steps

As with previous sizeable features, the design and implementation work will be carried out in a feature branch -- `feature/mempool-qos`, bifurcating from `main` -- to avoid the risk of last-minute serious problems. A feature branch enables the (management) decision to defer shipping a feature in a new release if it is considered not ready.

If we need our users to validate our design/implementation while this feature is in progress, we will probably need to maintain some temporary branch off a released version (likely `v0.38.x`) that reflects our work in `feature/mempool-qos`. We will work out the details as we start creating PRs with code.

#### Initial steps

- [x] Create branch `feature/mempool-qos`
- [x] Initial ADR skeleton with structure and discussions
  - #2868

#### Other related pending improvements on mempool

- [x] #2317
- [x] #895
- [ ] ~~Remove `TxsAvailable` and `CreateEmptyBlocks` (laundry list)~~ (not actually needed)

#### Design: core work

These tasks are mostly sequential, although some of them can be done in parallel

- [x] #2934
- [x] #3027
- [x] #3110
  - [2 days left]
- [x] #3088
- [x] ADR: Iterate on the design with output from users
- [x] #3138
  - [2 days left]
- [x] Decide where the design/spec will land [1 day]
  - Shall we keep the design/spec in the ADR? Or do we need to create a section in the spec?
  - We decided to keep it all in the ADR because it makes sense to understand the whole text. Later we will copy the Properties section to its own file in the spec folder.
- [x] Detailed implementation plan: flesh out tasks in section below
- [x] #3499
- [x] #4120
- [X] ~~Finalize spec/design document (ADR or move elsewhere) [7 days]~~
  - We decided to leave the current mini spec as part of the ADR.

Initial estimate (this section): 12 days

#### Implementation

- [x] #3084
- [x] #3303
- [x] #3296
  - ~~Implemented in Dog's code: Take it from there~~
- [X] Post the rest of the code of current prototype implementation in one or more PRs (in feature branch) [1 day]
  - Done in branch [hvanz/mempool-lanes-prototype](https://github.com/cometbft/cometbft/compare/main...hvanz/mempool-lanes-prototype)
- [x] #3502 
   - [1 day]
- [x] #3503 
   - [3 days]
- [x] #3475 
   - [6 days]
  - Move `Info` out of Handshake (likely) to `NewNode` 
- [x] #3484
  - [6 days]
  - This will change the internal data structures of CListMempool. It will be difficult to split it in multiple tasks/PRs.
  - Probably the prioritisation algorithm (WRR) will also be part of this PR.
  - [x] Implement lanes as multiple CLists
  - [x] Implement _WRR_ scheduling algorithm [5 days]
    - Classical WRR is implemented in the prototype, though only tested on the surface.
  - [x] Basic testing
- [x] #3506 
   - [1 day]
- [x] #3974
   - [4 days]
- [x] #3977 
   - [8 days]
- [x] #4005
  - #3825 
- [x] #4011
- [x] #4012
- [x] #4014
- [x] #4043
- [x] #4049
  - [7 days]

Initial estimate (this section): 50 days

#### Final steps

- [x] #4168 
- [x] Merge `feature/mempool-qos` back into `main` when the implementation is validated (200 node run, without report)
  - [2 days]
  - #4178
  - #4176
- [X] Figure out how to expose `feature/mempool-qos` on top of `v0.38.x`, `v0.38.x-experimental`, or `v1.x-experimental` [1 day]
  - #4179 
  - [X] ~~Deal with the work done off `v0.38.x` (merge or squash it into `v0.38.x-experimental`) [7 days]~~
  - We have backported Lanes to `v1.x-experimental`. Cosmos SDK will integrate Lanes from that branch (https://github.com/cosmos/cosmos-sdk/issues/21780). For the moment we won't backport to `v0.38.x` or `v0.38.x-experimental` unless there are interested users especially requiring these versions.
- [x] #4138
- [X] Revisit the existing mempool spec (by @hvanz and @otrack): is it compatible with mempool QoS? Do we need to work on it? [2 days]
  - There are three specs (not in this repo): one [high-level in English](https://github.com/cometbft/knowledge-base/blob/main/protocols/mempool-overview.md), one [high-level in Quint](https://github.com/cometbft/knowledge-base/blob/main/quint/mempool/Mempool.qnt), and one [low-level in Quint](https://github.com/cometbft/knowledge-base/tree/main/quint/mempool/v0). The two high-level specs are compatible with Lanes. The low-level spec follows closely the v0.38 code because the original goal was to use it to apply MBT. This last spec would need to be updated to be compatible with lanes.

Initial estimate (this section): 12 days

Additional changes after merging to main:
- #4229
- #4175 (required for sending tx load on DigitalOcean testnets)
- #4181
- #4241

**Comment by adizere** [comment]
Please see https://github.com/cosmos/ibc-go/issues/6159 while doing the research here. Glad to provide IBC-side related context of background if that would help.

**Comment by hdevalence** [comment]
This is an interesting proposal. I wanted to flag, however, that it doesn't serve Penumbra's needs, and while I can't speak on behalf of other users I am unconvinced that it solves the problem in general either.

Specifically, while the proposal allows an application to specify a few statically-designated lanes, and assign transactions to those lanes, it provides no way to prioritize transactions WITHIN those lanes. But this is the actual problem faced by an application receiving more transactions than there is execution capacity — which is the context where this problem needs to be solved in the first place.

Penumbra is designed so that every transaction has a chain-determined base fee and a user-specified transaction fee. The transaction fee minus the base fee is the proposer tip, and transactions should be priority ordered by the proposer tip. This is not possible to implement with the proposed design, which only allows defining a small number (as the number linearly replicates internal comet data structures) of lanes and assigning transactions to them, but does not allow specifying a transaction ordering within each lane. How is the app supposed to assign transactions to lanes to respect fee priority order? It cannot create one lane for every possible fee value, and it cannot statically quantize fee amounts into "priority groups" because the distribution of fee amounts is not known a priori.

The notes describe a FCFS ordering as a desirable goal. This is both a conceptual and practical mistake. Conceptually, FCFS is not a transaction ordering, because different nodes may receive transactions in different orders. An ordering must be a property of **a set of transactions**, i.e., using information in the transactions themselves. Practically, it is also a mistake because FCFS allows abusive users to spam low value transactions without consequence.   For example, while this proposal would allow an app to define an IBC lane to prioritize IBC packets, this would do absolutely nothing about the IBC spam currently ongoing on deployed chains, because it does not allow any prioritization within a lane. In other words, it is **only useful for enshrined, permissioned transaction submission, and not useful for transactions made by users of the chain.**

Every other blockchain that has users and therefore experiences congestion has some kind of fee market (emergent or otherwise) that allows transactions to be sorted by fee. Comet should do the same. At the very least, this should be described in the alternatives section.

**Comment by hvanz** [comment]
Thanks for the feedback @hdevalence! I think what you want is something like the deprecated [priority mempool](https://github.com/cometbft/cometbft/blob/hvanz/mempool-qos-technical-design/docs/references/architecture/adr-113-mempool-lanes.md#cometbft-priority-mempool), where each tx has a priority? One of the main requirements for this proposal was to respect the (best-effort) FIFO ordering because it's needed by IBC, so in this initial design all lanes will use the same scheduling algorithm for transaction dissemination and for reaping. That doesn't mean that in the future we can't change how each lane prioritize its transactions by customizing the scheduling logic per lane, probably with a callback defined in the application.

**Comment by adizere** [comment]
Really appreciate the insights @hdevalence, in particular the perspective from Penumbra which helps understand better the application requirements.

I wonder if the concern about within-lane prioritization (or roughly equivalent concern of fee market) goes into the direction of https://github.com/cometbft/cometbft/discussions/1112. If so, would love to have your input on that discussion; if not then I'll open another issue dedicated to within-lane prio work to gather requirements.
